File: app/src/main/java/com/artemchep/horario/_new/fragments/SubjectStreamFragment.java
package com.artemchep.horario._new.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.artemchep.horario.R;
import com.artemchep.horario._new.activities.SubjectActivity;
import com.artemchep.horario._new.adapters.AdapterModel;
import com.artemchep.horario._new.adapters.StreamAdapter;
import com.artemchep.horario.aggregator.Filter;
import com.artemchep.horario.database.models.SubjectInfo;
import com.artemchep.horario.database.models.SubjectTask;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;

import java.util.Comparator;
import java.util.List;

/**
 * @author Artem Chepurnoy
 */
public class SubjectStreamFragment extends FragmentModel<SubjectTask> implements SubjectActivity.Page {

    private ChildEventListener mChildEventListener = new ChildEventListener() {
        @Override
        public void onChildAdded(DataSnapshot dataSnapshot, String s) {
            SubjectTask task = dataSnapshot.getValue(SubjectTask.class);
            task.key = dataSnapshot.getKey();
            getAggregator().put(task);
        }

        @Override
        public void onChildChanged(DataSnapshot dataSnapshot, String s) {
            SubjectTask task = dataSnapshot.getValue(SubjectTask.class);
            task.key = dataSnapshot.getKey();
            getAggregator().put(task);
        }

        @Override
        public void onChildRemoved(DataSnapshot dataSnapshot) {
            getAggregator().remove(dataSnapshot.getKey());
        }

        @Override
        public void onChildMoved(DataSnapshot dataSnapshot, String s) {
        }

        @Override
        public void onCancelled(DatabaseError databaseError) {
        }
    };

    @Nullable
    @Override
    public View onCreateView(
            LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.___fragment_subject_stream, container, false);
    }

    @Override
    public void onStart() {
        super.onStart();
        getAggregator().reset();
        FirebaseDatabase.getInstance().getReference()
                .child("subjects").child("testing_is_fun").child("tasks")
                .addChildEventListener(mChildEventListener);
    }

    @Override
    public void onStop() {
        super.onStop();
        FirebaseDatabase.getInstance().getReference()
                .child("subjects").child("testing_is_fun").child("tasks")
                .removeEventListener(mChildEventListener);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setSubject(@Nullable SubjectInfo subject) {
    }

    @Override
    public Filter<SubjectTask> onCreateFilter() {
        return new Filter<>();
    }

    @Override
    public Comparator<SubjectTask> onCreateComparator() {
        return new Comparator<SubjectTask>() {
            @Override
            public int compare(SubjectTask a, SubjectTask b) {
                return a.key.compareToIgnoreCase(b.key);
            }
        };
    }

    @Override
    public AdapterModel<SubjectTask, ? extends AdapterModel.ViewHolder<SubjectTask>> onCreateAdapter(List<SubjectTask> list) {
        return new StreamAdapter(this, list);
    }

    @Override
    public void onFabClick(@NonNull View view) {
    }

    @Override
    public boolean hasFab() {
        return true;
    }

}

File: app/src/main/java/com/artemchep/horario/ui/fragments/SubjectStreamFragment.kt
package com.artemchep.horario.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.artemchep.horario.Palette
import com.artemchep.horario.R
import com.artemchep.horario._new.activities.SubjectActivity
import com.artemchep.horario.database.data.Post
import com.artemchep.horario.database.models.IPost
import com.artemchep.horario.extensions.inverted
import com.artemchep.horario.extensions.setTextExclusive
import com.artemchep.horario.interfaces.FragmentHost
import com.artemchep.horario.ui.adapters.AdapterBase
import com.artemchep.horario.ui.utils.*
import com.artemchep.horario.ui.widgets.PrettyTimeView
import com.artemchep.horario.ui.widgets.UserView
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import org.sufficientlysecure.htmltextview.HtmlTextView
import java.util.*

/**
 * @author Artem Chepurnoy
 */
class SubjectStreamFragment : FragmentStore<IPost>(), AdapterBase.OnItemClickListener, SubjectActivity.Page {

    private var color: Int = 0
    private lateinit var userId: String
    private lateinit var subjectId: String

    /** Reference to a list of the posts of this subject */
    private lateinit var postsQuery: Query

    override val filter: (IPost) -> Boolean = { true }
    override val comparator: Comparator<IPost> = compareBy<IPost>(
            { it.timestamp },
            { it.id }
    ).inverted()

    override fun onAttach(context: Context?) {
        arguments!!.run {
            userId = getString(EXTRA_USER_ID)
            subjectId = getString(EXTRA_SUBJECT_ID)
            color = getInt(EXTRA_COLOR, Palette.UNKNOWN)
        }

        val firestore = FirebaseFirestore.getInstance()
        postsQuery = firestore.collection("subjects/$subjectId/posts")
        super.onAttach(context)
    }

    override fun onCreateAdapter(list: MutableList<IPost>): Adapter = Adapter(context!!, list)

    override fun onCreateContentView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.___fragment_subject_stream, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.findViewById<RecyclerView>(R.id.recycler).run {
            // Show two columns of posts in
            // `large` interfaces
            layoutManager = if (isLarge) {
                StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
            } else LinearLayoutManager(activity)
        }
    }

    override fun onItemClick(view: View, model: IPost) {
        super.onItemClick(view, model)

        val args = Bundle().apply {
            putInt(EXTRA_COLOR, color)
            putString(EXTRA_USER_ID, userId)
            putString(EXTRA_SUBJECT_ID, subjectId)
            putString(EXTRA_POST_ID, model.getKey())
            putParcelable(EXTRA_POST, model)
        }

        val host = activity as FragmentHost
        host.fragmentShow(SubjectPostFragment::class.java, args, FragmentHost.FLAG_AS_SECONDARY)
    }

    override fun onFabClick(view: View) {
        val args = Bundle().apply {
            putInt(EXTRA_COLOR, color)
            putString(EXTRA_USER_ID, userId)
            putString(EXTRA_SUBJECT_ID, subjectId)
        }

        val host = activity as FragmentHost
        host.fragmentShow(SubjectPostEditFragment::class.java, args, FragmentHost.FLAG_AS_SECONDARY)
    }

    override fun hasFab(): Int = R.drawable.ic_plus_white_24dp

    override fun getQuery(): Query = postsQuery

    override fun getModel(snapshot: DocumentSnapshot): IPost = Post.from(snapshot)!!

    /**
     * @author Artem Chepurnoy
     */
    class Adapter(private val context: Context, list: List<IPost>) : AdapterBase<IPost, Adapter.ViewHolder>(list) {

        /**
         * @author Artem Chepurnoy
         */
        class ViewHolder(adapter: Adapter, v: View) : AdapterBase.ViewHolder<Adapter>(adapter, v) {
            internal val commentTextView: TextView = v.findViewById(R.id.info)
            internal val titleTextView: TextView = v.findViewById(R.id.title)
            internal val typeTextView: TextView = v.findViewById(R.id.type)
            internal val textTextView: HtmlTextView = v.findViewById(R.id.text)
            internal val timeView: PrettyTimeView = v.findViewById(R.id.time)
            internal val userView: UserView = v.findViewById(R.id.user_view)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            val v = inflater.inflate(R.layout.___item_post, parent, false)
            return ViewHolder(this, v)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val post = getItem(position)
            val res = context.resources

            if (post.text != null) {
                holder.textTextView.setHtml(post.text!!)
                holder.textTextView.movementMethod = null
            } else holder.textTextView.text = null

            holder.timeView.setTime(post.timestamp)
            holder.userView.modelId = post.author

            holder.typeTextView.setTextExclusive(when (post.type) {
                Post.TYPE_ANNOUNCEMENT -> res.getString(R.string.announcement)
                Post.TYPE_QUESTION -> res.getString(R.string.question)
                else -> null
            })

            holder.titleTextView.setTextExclusive(post.title)
            holder.commentTextView.setTextExclusive(post.commentsCounter
                    .takeIf { post.commentsCounter > 0 }
                    ?.let { res.getQuantityString(R.plurals.comments_counter, it, it) }
            )
        }

    }

}
